scripts/rpc-maas-tool.py

```diff
@@ -13,199 +13,600 @@
+                    alarm.entity = entity
+                    check.alarms.append(alarm)
+                    entity.alarms.append(alarm)
+
+    def _add_metrics_list_to_check(self, check):
+        """Called via ThreadPoolExecutor, result returned via queue."""
+        metrics = self.conn.list_metrics(check.entity.id, check.id)
+        self.q.put((check, metrics))
+
+    def add_metrics(self):
+        """Add metrics list to each checks
+
+        Requires a call per check, so ThreadPoolExecutor is used to
+        parallelise and reduce time taken
+        """
+        with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:
+            for entity in self.get_entities():
+                for check in entity.checks:
+                    executor.submit(self._add_metrics_list_to_check, check)
+        while not self.q.empty():
+            check, metrics = self.q.get()
+            for metric in metrics:
+                metric.check = check
+                metric.entity = check.entity
+                check.metrics.append(metric)
+                check.entity.metrics.append(metric)
+
+    def get_entities(self):
+        """Return list of known entities
+
+        entity_prefix filter not required as this is done in __init__
+        """
+        return self.entities
+
+    def get_checks(self, check_prefix=''):
+        """List checks for entities matching a prefix"""
+
+        checks = []
+        for entity in self.entities:
+            checks.extend([c for c in entity.checks
+                           if check_prefix in c.label])
+        return checks
+
+    def get_alarms(self, entry, check):
+        """Get list of alarms
+
+        Params:
+            entry: overview dictionary for one entity.
+
+        This function adds a state field to each alarm object
+        using information form the 'latest_alarm_states' entry key.
+        """
+        alarms = []
+        for alarm in entry['alarms']:
+            if not alarm.check_id == check.id:
+                continue
+            # add state to the alarm object from latest_alarm_states
+            alarm_states = sorted(
+                [als for als in entry['latest_alarm_states']
+                    if als.alarm_id == alarm.id],
+                key=lambda x: x.timestamp
+            )
+            if alarm_states:
+                alarm.state = alarm_states[-1].state
+            else:
+                alarm.state = "UNKNOWN"
+            alarms.append(alarm)
+        return alarms
+
+
+class RpcMaasAgentConfig(object):
+    """Read MAAS Agent configuration files
+
+    Parse them as yaml and store that.
+    """
+    def __init__(self, agentconfdpath):
+        self.agentconfdpath = agentconfdpath
+        self.checks = self._parse_maas_configs()
+
+    def _parse_maas_configs(self):
+        """Read all config files in agentconfdpath"""
+        self.checks = {}
+        for path in list(os.walk(self.agentconfdpath))[0][2]:
```

Comment:
'list(os.walk(self.agentconfdpath))[0][2]' is hard to read, at least for me. Why not use os.listdir instead? The directory shouldn't contain any sub-directories but it you're worried you could always catch the exceptions.